src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs

```diff
@@ -485,107 +473,139 @@ private unsafe void GenerateStatus()
-                        return;
-                    }
+                _name = userGivenName;
+                _displayName = result;
+                _eitherName = null;
+            }
+            else if (String.IsNullOrEmpty(_displayName))
+            {
+                // We must have _name
+                string result = GetServiceDisplayName(_serviceManagerHandle, _name);
+
+                if (result == null)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.NoService, _name, _machineName), new Win32Exception(Interop.Errors.ERROR_SERVICE_DOES_NOT_EXIST));
                 }
 
-                throw new InvalidOperationException(SR.Format(SR.NoService, _eitherName, _machineName));
+                _displayName = result;
+                _eitherName = null;
             }
-            finally
+        }
+
+        /// <summary>
+        /// Gets service name (key name) from service display name.
+        /// Returns null if service is not found.
+        /// </summary>
+        private unsafe string GetServiceKeyName(SafeServiceHandle SCMHandle, string serviceDisplayName)
+        {
+            Span<char> initialBuffer = stackalloc char[256];
+            var builder = new ValueStringBuilder(initialBuffer);
+            int bufLen;
+            while (true)
             {
-                Marshal.FreeHGlobal(memory);
-                if (databaseHandle != IntPtr.Zero)
+                bufLen = builder.Capacity;
+                fixed (char* c = &builder.GetPinnableReference())
+                {
+                    if (Interop.Advapi32.GetServiceKeyName(SCMHandle, serviceDisplayName, c, ref bufLen))
+                        break;
+                }
+
+                int lastError = Marshal.GetLastWin32Error();
+                if (lastError == Interop.Errors.ERROR_SERVICE_DOES_NOT_EXIST)
+                {
+                    return null;
+                }
+                else if (lastError != Interop.Errors.ERROR_INSUFFICIENT_BUFFER)
                 {
-                    Interop.Advapi32.CloseServiceHandle(databaseHandle);
+                    throw new InvalidOperationException(SR.Format(SR.NoService, serviceDisplayName, _machineName), new Win32Exception(lastError));
                 }
+
+                builder.EnsureCapacity(bufLen + 1); // Does not include null
             }
+
+            builder.Length = bufLen;
+            return builder.ToString();
         }
 
-        private static IntPtr GetDataBaseHandleWithAccess(string machineName, int serviceControlManagerAccess)
+        private unsafe string GetServiceDisplayName(SafeServiceHandle SCMHandle, string serviceName)
         {
-            IntPtr databaseHandle = IntPtr.Zero;
+            var builder = new ValueStringBuilder(4096);
+            int bufLen;
+            while (true)
+            {
+                bufLen = builder.Capacity;
+                fixed (char* c = &builder.GetPinnableReference())
```

Comment:
I think this can be just `fixed (char* c = builder)` with the latest C# compiler.

Reply:
I had to add a parameterless overload to ValueStringBuilder to achieve this useful trick.